### DISCUSSION TITLE
Unsuccessful installation, Portable Win32 version

### DISCUSSION THREAD

**Discussion by tumatanquang** [WinMerge/winmerge#1708] [Q&A] [answered]
Hello, I am using the [WinMerge-2.16.26-x64-PerUser-Setup.exe](https://github.com/WinMerge/winmerge/releases/download/v2.16.26/WinMerge-2.16.26-x64-PerUser-Setup.exe) file to install. But in `Preparing to Install` step, after clicking `Next`, I received an error:
> Internal error: An attempt was made to expand the "group" constant before it was initialized.
![image](https://user-images.githubusercontent.com/20223064/217573510-2e437a6b-f91e-412a-80a2-35fe13ca972b.png)

I tried deleting and re-downloading the installation file a few times but still got that error.
How to solve this error?
In addition, I want to know if there is a Portable version (no need to install) for Win32 (x86)? I checked on the [SourceForge download page](https://sourceforge.net/projects/winmerge/files/stable/2.16.8/), it seems that only the Win32 setup version for all users (without PerUser and portable version)?
Thank you.

**Comment by sdottaka** [comment, marked as answer]
You seem to use Windows 7, but I could not reproduce this on my Windows 7 (64bit) PC. Are you installing via some unusual or special software?

> In addition, I want to know if there is a Portable version (no need to install) for Win32 (x86)?

The x86zip version, which does not require installation, can be downloaded from

https://github.com/WinMerge/winmerge/releases/download/v2.16.26/winmerge-2.16.26-exe.zip

**Reply by tumatanquang** [reply]
> You seem to use Windows 7, but I could not reproduce this on my Windows 7 (64bit) PC. Are you installing via some unusual or special software?

No, I install directly through the installation file. However, on the `Select Components` page, I selected `Compact Installation` and at the `Select Start Menu Folder` page, I selected `Don't create a Start Menu folder` (as shown in the attached image).
![image](https://user-images.githubusercontent.com/20223064/224496842-c55ef2d6-d5fa-4c3b-bb12-b9f96d905ea2.png)
![image](https://user-images.githubusercontent.com/20223064/224496675-4fe81c7e-0310-4ec7-8e7c-61f2a0c6812e.png)

**Reply by sdottaka** [reply]
Thanks for the information. I have tried the following, but could not reproduce the problem. I would appreciate it if you could let me know if there are any other conditions to reproduce the problem.

- Create a new non-administrator user in Windows 7.
- Run the installer WinMerge-2.16.26-x64-PerUser-Setup.exe as that user.
- Select Compact Installation and check `Don't create a Start Menu folder

**Reply by tumatanquang** [reply]
> Thanks for the information. I have tried the following, but could not reproduce the problem. I would appreciate it if you could let me know if there are any other conditions to reproduce the problem.
> 
> * Create a new non-administrator user in Windows 7.
> * Run the installer WinMerge-2.16.26-x64-PerUser-Setup.exe as that user.
> * Select Compact Installation and check `Don't create a Start Menu folder

Hi, I tried and still got the same error. I think it will be easier to determine the cause of the error if adding an error code and an error message if the error continues to occur in the next update version.

**Reply by sdottaka** [reply]
I was able to reproduce this issue.

  The text box under "To continue, click Next. If you would like to select a different folder, click Browse" on the "Select Start Menu Folder" page is initially set to "WinMerge". If this is empty as follows, this message will be displayed.
![image](https://user-images.githubusercontent.com/98126/226108303-23427bbe-c31a-41a2-8a42-b0305d64ae6f.png)
![image](https://user-images.githubusercontent.com/98126/226108308-14f8d72a-f4fb-4beb-8425-2c44a49b4330.png)

I don't think the error will occur if this textbox is not empty. Is this textbox empty from the beginning on your PC?

In addition, I have tried changing the installer script in various ways so that an error does not occur when this is empty, but I have not been able to avoid the problem.

**Reply by tumatanquang** [reply]
> I don't think the error will occur if this textbox is not empty.

That's right, I tried to reproduce the problem, if the text box is not empty then no error will appear.

> Is this textbox empty from the beginning on your PC?

Probably not. Perhaps at that time I wanted to change to another folder in the Browse window, but then I changed my mind and closed the browsing window and ticked into the `Don't create a Start Menu folder`.

> In addition, I have tried changing the installer script in various ways so that an error does not occur when this is empty, but I have not been able to avoid the problem.

I don't know what the installer creation tool/script of this software is but I have tried to reproduce this issue on a lot of other software that has an installer interface like this one, but what those software have in common after I tried to reproduce this issue is: There are no errors if you check `Don't create a Start Menu folder`, whether the text box is empty or not! I even tried deleting the entire path in the textbox, but simply checking `Don't create a Start Menu folder` won't give an error.
I think the tool/script that this program used to create the installer is still doing a path check in the text box and trying to create a shortcut for the program and then checking if the user ticks `Don't create a Start Menu folder` or not. If so, delete the previously created shortcut.
That's the assumption I made after trying a lot of other software that has a similar installer interface but doesn't get an error whether the textbox is empty or not.

**Reply by sdottaka** [reply]
Thank you for reminding me. Upon retrying, I was able to fix it (with commit ad82e856784af137431ea124c4586c731ea3dbb4). 
The next release should not have this issue anymore.

**Reply by tumatanquang** [reply]
> Thank you for reminding me. Upon retrying, I was able to fix it (with commit [ad82e85](https://github.com/WinMerge/winmerge/commit/ad82e856784af137431ea124c4586c731ea3dbb4)). The next release should not have this issue anymore.

It's so amazing!
